File: php/42-trapping-rain-water.php
<?php
class Solution {

    /**
     * @param Integer[] $height
     * @return Integer
     */
    function trap($height) {
        $count = count($height);
        if ($count <= 2) {
            return 0;
        }
        //单调递减栈
        $stack = [];
        $water = 0;
        for ($i = 0; $i < $count; $i++) {
            $top = end($stack);
            if (empty($stack) || $height[$top] > $height[$i]) {
                array_push($stack, $i);
            } else {
                while (!empty($stack) && $height[end($stack)] < $height[$i]) {
                    $v = array_pop($stack);
                    if (!empty($stack)) {
                        $left = end($stack);
                        $w = $i - $left - 1;
                        $h = min($height[$left], $height[$i]) - $height[$v];
                        $water += $w * $h;
                    }
                }
                array_push($stack, $i);
            }
        }
        return $water;
    }
}

File: go/42-trapping-rain-water.go
package main

func trap(height []int) int {
	count := len(height)
	if count <= 2 {
		return 0
	}
	water := 0
	stack := make([]int, 0)
	for i, v := range height {
		if len(stack) == 0 || height[stack[len(stack)-1]] > v {
			stack = append(stack, i)
		} else {
			for len(stack) > 0 && height[stack[len(stack)-1]] < v {
				b := stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				if len(stack) > 0 {
					left := stack[len(stack)-1]
					w := i - left - 1
					h := int(math.Min(float64(height[i]), float64(height[left]))) - height[b]
					water += w * h
				}
			}
			stack = append(stack, i)
		}
	}
	return water
}
